proposals/explicit-api-mode.md

```diff
@@ -0,0 +1,154 @@
+
+### Experimental API
+
+When one is writing a library, they should use `@Experimental(...)` annotation to propagate the experimental status of types they use.
+If experimental types are used as implementation details across all library, it might be convenient to mark the whole module with `-Xuse-experimental`.
+In that case, it would be easy to forget to mark the corresponding public API as propagative.
+Therefore, in Explicit API mode, the compiler would still require explicit `@Experimental` or `@UseExperimental` annotation on a declaration with experimental types in the signature, even if the whole module accepts experimental status via `-Xuse-experimental`.
+
+### Inspection exclusions
+
+After careful review, we decided that some declarations should not require explicit visibility modifiers even in 'Explicit API' mode for the sake of readability, simplicity, and common sense. These declarations are:
+
+1. Primary constructors
+
+    Because explicit visibility also requires you to insert keyword `constructor`.
+
+2. Properties of data classes
+
+    Because such classes usually fit in one line of code and do not profit much from information hiding.
+
+3. Methods marked with `override`
+
+    Because default visibility for such methods is the visibility of an overridden method.
+
+4. Property getters and setters
+
+    Because getters can't change visibility and setter-only explicit visibility looks ugly.
+
+However, if you still want to insert an explicit visibility modifier for such declarations, it would not be marked as redundant by IDE.
+
+## Implementation
+
+This mode is enabled by the compiler flag `-Xexplicit-api={strict|warning}`. `strict` state of the flag means that compiler will issue errors when public API declaration does not have explicit visibility or explicit return type.
+`warning` means that the compiler will issue warnings (this will help migration).
+Note: missing KDoc is always a warning, regardless of the state of the flag.
+
+To ease setting up this mode, a DSL would be provided in Kotlin Gradle plugin:
+
+```gradle
+kotlinOptions {
+    explicitApi()
+    // or
+    explicitApi = ExplicitApiMode.Strict
+    // or
+    explicitApiWarning()
+    // or
+    explicitApi = ExplicitApiMode.Warning
+}
+```
+Explicit mode enabled here would not affect test sources.
+
+The exact place of these methods (top-level `kotlinOptions`, or compilation, or particular source set) is TBA. Maven plugin option is TBD.
+
+Kotlin plugin in IntelliJ IDEA would recognize that explicit mode is enabled and would offer corresponding intentions and quick-fixes.
+
+The following **IDE** inspections would be disabled either because they're redundant or duplicated in the compiler:
+
+* 'Redundant visibility modifier' for public API declarations
+* 'Public API declaration has implicit return type' (would be replaced with compiler diagnostic)
```

Comment:
It would be good to relax this restriction for `const val` types, as they are only used for primitives and `String`. I created [an issue](https://youtrack.jetbrains.com/issue/KT-30512) about this topic some time ago.